### ISSUE TITLE
spdk_nvme_ctrlr_alloc_io_qpair() failed 

### ISSUE THREAD

**Issue by flaging** [spdk/spdk#2078]
## Expected Behavior

I want to bind one thread with multi-nvmes, so I create multi-queue for one thread. However, when the second create second thread do the same thing, It comes with a SIGSEGV.

## Current Behavior

![image](https://user-images.githubusercontent.com/10471450/128672856-2c01ff97-fb0d-467e-91ad-a8af3f666937.png)

## Context (Environment including OS version, SPDK version, etc.)

SPDK: v17.07
OS: CentOS 7 
kernel: 5.8.0-dfl

Is there any limitation for interface spdk_nvme_ctrlr_alloc_io_qpair() ? Or one ctrlr can only bind with one qpair??

**Comment by tomzawadzki** [comment]
>SPDK: v17.07
Can you please try a recent SPDK ? Latest master or one of the supported releases SPDK 21.07 or SPDK 21.01 LTS.

**Comment by flaging** [comment]
> > SPDK: v17.07
> 
> Can you please try a recent SPDK ? Latest master or one of the supported releases SPDK 21.07 or SPDK 21.01 LTS.

I have changed my test server and this disappeared, Thanks anyway. 

I have some change in v17.07 like register device's bar space to the spdk's map. I can't get the virtual addr of this space, So    I bind some unused virtual addr and the device's phy addr by force. In newer version like v21.07, I cant find the same interface to bind both virtual addr and phy addr, or register the device's space to spdk automatically. Any advices ? Thanks in advance!

**Comment by jimharris** [comment]
Hi @flaging,

Can you explain more about this "binding by force"?  In SPDK, all PCI devices have an associated DPDK driver associated with them.  For example, see lib/env_dpdk/pci_ioat.c and then how lib/ioat/ioat.c uses spdk_pci_ioat_get_driver() in the call to spdk_pci_enumerate().

If you have a PCI device that you would like its BARs mapped into virtual address space, this is the supported way to do it.

Closing this issue since the original problem was resolved, but feel free to post more questions here or on Slack.

Thanks,

Jim

**Comment by flaging** [comment]
Thanks for your advice!! 

I have checked my device, which could not support IOAT.  Maybe I could not do as i want.

The implement "bind by force" is that, I mmap some virtual address, and get the phy address as well.  Then i create one function, in which I can register virtual addr and phy addr, not just register virtual addr and length.

Then in the spdk namespace, when i execute some thing with virtual addr, the spdk does it with my phy address. It is far more easy in spdk v17.07, for i can see the phy addr in spdk_vtophys_register().

Not long after this version, this interface was merged to memory.c, and it is not easy to see the phy address easily. That's why i don't upgrade my spdk version.
